libdrgn/iterators.c

```diff
@@ -0,0 +1,486 @@
+						    frame->loop_index);
+			if (!err)
+				err = radix_tree_iter_handle_node(
+					iter, &node,
+					frame->index + (frame->loop_index++
+							<< frame->shift),
+					&entry_populated);
+		} else {
+			drgn_object_deinit(&frame->slots);
+			frames_pop(&iter->frames);
+		}
+	}
+	*is_exhausted = !entry_populated;
+	drgn_object_deinit(&node);
+	return err;
+}
+
+struct drgn_error *radix_tree_iter_init(struct radix_tree_iter *iter,
+					const struct drgn_object *root)
+{
+	struct drgn_program *prog = drgn_object_program(root);
+	iter->started = false;
+	drgn_object_init(&iter->root, prog);
+	drgn_object_init(&iter->entry.node, prog);
+	iter->entry.index = 0;
+	iter->prog = prog;
+	frames_init(&iter->frames);
+
+	struct drgn_error *err;
+	struct drgn_object tmp;
+
+	drgn_object_init(&tmp, prog);
+
+	/* node = root->xa_head */
+	err = drgn_object_member(&iter->root, root, "xa_head");
+	if (!err) {
+		err = drgn_program_find_type(prog, "struct xa_node *", NULL,
+					     &iter->node_type);
+		if (err)
+			goto out;
+		iter->RADIX_TREE_INTERNAL_NODE = 2;
+	} else if (err->code == DRGN_ERROR_LOOKUP) {
+		drgn_error_destroy(err);
+		/* node = (void *)root.rnode */
+		err = drgn_object_member_dereference(&iter->root, root,
+						     "rnode");
+		if (err)
+			goto out;
+		err = drgn_program_find_type(prog, "void *", NULL,
+					     &iter->node_type);
+		if (err)
+			goto out;
+		err = drgn_object_cast(&iter->root, iter->node_type,
+				       &iter->root);
+		if (err)
+			goto out;
+		err = drgn_program_find_type(prog, "struct radix_tree_node *",
+					     NULL, &iter->node_type);
+		if (err)
+			goto out;
+		iter->RADIX_TREE_INTERNAL_NODE = 1;
+	} else {
+		goto out;
+	}
+
+	struct drgn_type_member *member;
+	uint64_t member_bit_offset;
+	err = drgn_type_find_member(drgn_type_type(iter->node_type.type).type,
+				    "slots", &member, &member_bit_offset);
+	if (err)
+		goto out;
+	struct drgn_qualified_type member_type;
+	err = drgn_member_type(member, &member_type, NULL);
+	if (err)
+		goto out;
+	if (drgn_type_kind(member_type.type) != DRGN_TYPE_ARRAY) {
+		err = drgn_error_create(
+			DRGN_ERROR_TYPE,
+			"struct radix_tree_node slots member is not an array");
+		goto out;
+	}
+	iter->RADIX_TREE_MAP_MASK = drgn_type_length(member_type.type) - 1;
+
+out:
+	if (err) {
+		drgn_object_deinit(&iter->entry.node);
+		frames_deinit(&iter->frames);
+	}
+	drgn_object_deinit(&tmp);
+	return err;
+}
+
+void radix_tree_iter_deinit(struct radix_tree_iter *iter)
+{
+	drgn_object_deinit(&iter->root);
+	drgn_object_deinit(&iter->entry.node);
+	while (iter->frames.size) {
+		drgn_object_deinit(&frames_pop(&iter->frames)->slots);
+	}
+	frames_deinit(&iter->frames);
+}
+
+struct drgn_error *radix_tree_iter_next(struct radix_tree_iter *iter,
+					struct radix_tree_iter_entry **ret)
+{
+	bool is_exhausted = false;
+	bool entry_populated = false;
+	bool node_is_truthy = false;
+	struct drgn_error *err;
+	do {
+		err = iter->started ?
```

Comment:
This is a lot better, thanks. I think the only real difference between your approach and my Python one is how to handle the root node. Either way, it ends up being a special case.

A couple of further comments:

* Do you need this `iter->started` check on every iteration? Can't you handle the `!iter->started` case once before the loop? The two cases (handle the root node and handle the slots of an inner node) are different enough that I don't think it makes sense to handle them in the same place, even if they have some minor overlap.
* If you're able to move the `!iter->started` case out of the loop, then I think it'd be clearer to get rid of `radix_tree_iter_run_frame` and just do that inner loop in this function, too.

Reply:
Good points. I moved `radix_tree_iter_run_frame` into the inner loop, which became the _only_ loop as I realized by tweaking `radix_tree_iter_handle_node`'s behavior regarding null non-internal nodes, I could remove the outer loop altogether.